arch/arm/core/aarch64/arm_mmu.c

```diff
@@ -0,0 +1,464 @@
+#include <arch/arm/aarch64/cpu.h>
+#include <arch/arm/aarch64/arm_mmu.h>
+#include <linker/linker-defs.h>
+#include <sys/util.h>
+
+/* Set below flag to get debug prints */
+#define MMU_DEBUG_PRINTS	0
+/* To get prints from MMU driver, it has to initialized after console driver */
+#define MMU_DEBUG_PRIORITY	70
+
+#if MMU_DEBUG_PRINTS
+/* To dump page table entries while filling them, set DUMP_PTE macro */
+#define DUMP_PTE		0
+#define MMU_DEBUG(fmt, ...)	printk(fmt, ##__VA_ARGS__)
+#else
+#define MMU_DEBUG(...)
+#endif
+
+/* We support only 4kB translation granule */
+#define PAGE_SIZE_SHIFT		12U
+#define PAGE_SIZE		(1U << PAGE_SIZE_SHIFT)
+#define XLAT_TABLE_SIZE_SHIFT   PAGE_SIZE_SHIFT /* Size of one complete table */
+#define XLAT_TABLE_SIZE		(1U << XLAT_TABLE_SIZE_SHIFT)
+
+#define XLAT_TABLE_ENTRY_SIZE_SHIFT	3U /* Each table entry is 8 bytes */
+#define XLAT_TABLE_LEVEL_MAX	3U
+
+#define XLAT_TABLE_ENTRIES_SHIFT \
+			(XLAT_TABLE_SIZE_SHIFT - XLAT_TABLE_ENTRY_SIZE_SHIFT)
+#define XLAT_TABLE_ENTRIES	(1U << XLAT_TABLE_ENTRIES_SHIFT)
+
+/* Address size covered by each entry at given translation table level */
+#define L3_XLAT_VA_SIZE_SHIFT	PAGE_SIZE_SHIFT
+#define L2_XLAT_VA_SIZE_SHIFT	\
+			(L3_XLAT_VA_SIZE_SHIFT + XLAT_TABLE_ENTRIES_SHIFT)
+#define L1_XLAT_VA_SIZE_SHIFT	\
+			(L2_XLAT_VA_SIZE_SHIFT + XLAT_TABLE_ENTRIES_SHIFT)
+#define L0_XLAT_VA_SIZE_SHIFT	\
+			(L1_XLAT_VA_SIZE_SHIFT + XLAT_TABLE_ENTRIES_SHIFT)
+
+#define LEVEL_TO_VA_SIZE_SHIFT(level) \
+				(PAGE_SIZE_SHIFT + (XLAT_TABLE_ENTRIES_SHIFT * \
+				(XLAT_TABLE_LEVEL_MAX - (level))))
+
+/* Virtual Address Index within given translation table level */
+#define XLAT_TABLE_VA_IDX(va_addr, level) \
+	((va_addr >> LEVEL_TO_VA_SIZE_SHIFT(level)) & 0x1ffULL)
+
+/*
+ * Calculate the initial translation table level from CONGIG_ARM64_VA_BITS
+ * For a 4 KB page size,
+ * (va_bits <= 21)	 - base level 3
+ * (22 <= va_bits <= 30) - base level 2
+ * (31 <= va_bits <= 39) - base level 1
+ * (40 <= va_bits <= 47) - base level 0
+ */
+#define GET_XLAT_TABLE_BASE_LEVEL(va_bits)	\
+	((va_bits > L0_XLAT_VA_SIZE_SHIFT)	\
+	? 0U					\
+	: (va_bits > L1_XLAT_VA_SIZE_SHIFT)	\
+	? 1U					\
+	: (va_bits > L2_XLAT_VA_SIZE_SHIFT)	\
+	? 2U : 3U)
+
+#define XLAT_TABLE_BASE_LEVEL	GET_XLAT_TABLE_BASE_LEVEL(CONFIG_ARM64_VA_BITS)
+
+#define GET_NUM_BASE_LEVEL_ENTRIES(va_bits)	\
+	(1U << (va_bits - LEVEL_TO_VA_SIZE_SHIFT(XLAT_TABLE_BASE_LEVEL)))
+
+#define NUM_BASE_LEVEL_ENTRIES	GET_NUM_BASE_LEVEL_ENTRIES(CONFIG_ARM64_VA_BITS)
+
+#if DUMP_PTE
+#define L0_SPACE ""
+#define L1_SPACE "  "
+#define L2_SPACE "    "
+#define L3_SPACE "      "
+#define XLAT_TABLE_LEVEL_SPACE(level)		\
+	(((level) == 0) ? L0_SPACE :		\
+	((level) == 1) ? L1_SPACE :		\
+	((level) == 2) ? L2_SPACE : L3_SPACE)
+#endif
+
+static u64_t base_xlat_table[NUM_BASE_LEVEL_ENTRIES]
+		__aligned(NUM_BASE_LEVEL_ENTRIES * sizeof(u64_t));
+
+static u64_t xlat_tables[CONFIG_MAX_XLAT_TABLES][XLAT_TABLE_ENTRIES]
+		__aligned(XLAT_TABLE_SIZE);
```

Comment:
Can we use `XLAT_TABLE_ENTRIES * sizeof(u64_t)` for the sake of consistency?

Reply:
I didn't get this. 

`CONFIG_MAX_XLAT_TABLES ` is required number of translation tables, each with `XLAT_TABLE_ENTRIES`(i.e. 512 entries), and each entry of 8 B, so `CONFIG_MAX_XLAT_TABLES  `with each translation table of (8* 512 = 4kB).

`__aligned(XLAT_TABLE_SIZE)` is used because `xlat_tables `have to be `XLAT_TABLE_SIZE`(4kB in our case) aligned.

Reply:
my suggestion was only about:
```
__aligned(XLAT_TABLES_ENTRIES * sizeof(u64_t));
```

Reply:
ok, got it.